File: aoc2017/day01/src/test/java/de/havox_design/aoc2017/day01/Day01Test.java
package de.havox_design.aoc2017.day01;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Day01Test {

    @SuppressWarnings("java:S2699")
    @Test
    void testMainClass() {
        MainClass.main(new String[0]);
    }

    @ParameterizedTest
    @MethodSource("getDataForPart1")
    void testPart1(String fileName, long expected) {
        Assertions.assertEquals(expected, InverseCaptcha.solvePart1(fileName));
    }

    private static Stream<Arguments> getDataForPart1() {
        return Stream.of(
                Arguments.of("day01Part1sample1.txt", 3L),
                Arguments.of("day01Part1sample2.txt", 4L),
                Arguments.of("day01Part1sample3.txt", 0L),
                Arguments.of("day01Part1sample4.txt", 9L)
        );
    }

    @ParameterizedTest
    @MethodSource("getDataForPart2")
    void testPart2(String fileName, long expected) {
        Assertions.assertEquals(expected, InverseCaptcha.solvePart2(fileName));
    }

    private static Stream<Arguments> getDataForPart2() {
        return Stream.of(
                Arguments.of("day01Part2sample1.txt", 6L),
                Arguments.of("day01Part2sample2.txt", 0L),
                Arguments.of("day01Part2sample3.txt", 4L),
                Arguments.of("day01Part2sample4.txt", 12L),
                Arguments.of("day01Part2sample5.txt", 4L)
        );
    }
}

File: aoc2023/src/test/kotlin/de/havox_design/aoc2023/day01/Day01Test.kt
package de.havox_design.aoc2023.day01

import de.havox_design.aoc2023.MainClass
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class Day01Test {
    @Test
    fun testMainClass() {
        MainClass.main(arrayOf("day01"))
    }

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart1")
    fun testSolvePart1(filename: String, expectedResult: Long) =
        Trebuchet(filename).solvePart1().shouldBe(expectedResult)

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart2")
    fun testSolvePart2(filename: String, expectedResult: Long) =
        Trebuchet(filename).solvePart2().shouldBe(expectedResult)

    companion object {
        @JvmStatic
        private fun getDataForTestSolvePart1(): Stream<Arguments> =
            Stream.of(
                Arguments.of("de/havox_design/aoc2023/day01/day01Part1sample1.txt", 12L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part1sample2.txt", 38L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part1sample3.txt", 15L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part1sample4.txt", 77L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part1sample5.txt", 142L)
            )

        @JvmStatic
        private fun getDataForTestSolvePart2(): Stream<Arguments> =
            Stream.of(
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample1.txt", 29L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample2.txt", 83L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample3.txt", 13L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample4.txt", 24L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample5.txt", 42L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample6.txt", 14L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample7.txt", 76L),
                Arguments.of("de/havox_design/aoc2023/day01/day01Part2sample8.txt", 281L)
            )
    }
}

private fun Long.shouldBe(expectation: Long) = assertEquals(expectation, this)
